### PR TITLE
dump memory snapshots

### PR DESCRIPTION
**Pull request by weifengpy**
Screenshot 2025-07-03 at 3 38 36 PM

load .pkl into https://docs.pytorch.org/memory_viz. I made some annotation. The memory trace follows a normal pattern
* memory increases during forward because of acitvations
* memory decreases during backward becuase of releasing acitvations
* memory peaks at loss.backward()

maybe consdier activation checkpointing to reduce peak memory

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -292,22 +296,17 @@ def causal_mask_mod(b, h, q_idx, kv_idx):
...
+            # and freezes chrome
+            # drag .pkl file to https://docs.pytorch.org/memory_viz
+            torch.cuda.memory._dump_snapshot(f"prof_memsnap_r{rank}.pkl")
         if step == 6:
             torch.cuda.cudart().cudaProfilerStop()
             prof.stop()
             prof.export_chrome_trace(f"prof_trace_r{rank}.json.gz")  # TODO: speedup gz
             prof.export_stacks(f"prof_stacks_cpu_r{rank}.txt")
-            torch.cuda.memory._dump_snapshot(f"prof_memsnap_r{rank}.pkl")
-            torch.cuda.memory._record_memory_history(enabled=None)
```
Comment (weifengpy) [inline main.py:301]: `torch.cuda.memory._record_memory_history` overhead is very minimal, espectially when cpu thread is ahead of cuda streams

File: main.py [hunk 2]
```diff
@@ -292,22 +296,17 @@ def causal_mask_mod(b, h, q_idx, kv_idx):
...
-            # Sometimes fails on gpu1...
-            try:
-                prof.export_memory_timeline(f"prof_memtl_r{rank}.html")
-                prof.export_memory_timeline(f"prof_memtl_r{rank}.json.gz")  # TODO: speedup gz
-                prof.export_memory_timeline(f"prof_memtl_r{rank}.raw.json.gz")
```
Comment (weifengpy) [inline main.py:306]: _dump_snapshot and https://docs.pytorch.org/memory_viz is what pytorch dev use these days. see snapshot

**Comment by lucasb-eyer** [comment]
Fantastic, thanks a lot for taking a look and helping!
